File: engine/core/tests/test_revision.cpp
// test_revision.cpp — the remask_lowconf revision path (DESIGN §5.2), the "model changes its mind"
// feature, ported from lab/tests/test_revision.py::TestRevisionEndToEnd. Backend-free: the
// position-deterministic FakeAdapter peaks every committed slot to ~0.995 confidence, so
// tau_revise=1.0 re-masks every committed token once (cap=1) — a hard stress test that the
// per-position cap still guarantees termination under maximally aggressive revision. Also pins the
// opt-in contract: with revision off, the loop is byte-identical to the committed goldens.
#include "cloze/generate.hpp"
#include "cloze/model.hpp"
#include "fake_adapter.hpp"

#include <cassert>
#include <cstdio>
#include <map>
#include <variant>
#include <vector>

using namespace cloze;

namespace {
const std::vector<int> PROMPT = {1, 2, 3, 4, 5};  // p = 5; generated region is positions 5..8
}  // namespace

int main() {
    // (a) revisions fire, respect the per-position cap, stay in the generated region, terminate.
    {
        FakeAdapter fake(16, /*eos=*/-1, /*eos_at=*/-1);
        const int mask = fake.config().mask_token_id;
        const int p = static_cast<int>(PROMPT.size());
        GenerateConfig cfg{/*max_new=*/4, /*steps=*/20, /*block_len=*/0, /*topk=*/-1};
        // tau_revise=1.0 re-masks every committed token once (cap=1): maximal churn, must still drain.
        ReviseConfig revise{/*enabled=*/true, /*tau_revise=*/1.0, /*max_revisions=*/1};
        GenerateResult r = generate(fake, PROMPT, cfg, CacheConfig{}, nullptr, {}, revise);

        int n_revised_events = 0;
        std::map<int, int> per_pos;
        for (const Event& e : r.events) {
            if (const auto* tr = std::get_if<TokensRevised>(&e)) {
                ++n_revised_events;
                for (const ReviseItem& it : tr->items) {
                    assert(it.old != mask);            // a real, previously-committed token was retracted
                    assert(it.conf < 1.0);             // below tau_revise
                    assert(p <= it.pos && it.pos < p + cfg.max_new);  // generated region only
                    per_pos[it.pos] += 1;
                }
            }
        }
        assert(n_revised_events > 0);                  // revision actually happened
        for (const auto& kv : per_pos) assert(kv.second <= revise.max_revisions);  // cap honored
        assert(std::holds_alternative<GenFinished>(r.events.back()));  // it terminated
        assert(r.reason == "length");                  // and fully drained despite the churn
        assert((r.generated == std::vector<int>{5, 6, 7, 8}));  // same final board as the no-revise run
    }

    // (b) opt-in: revision off => no tokens_revised events and the byte-identical committed board.
    {
        FakeAdapter fake(16, -1, -1);
        GenerateConfig cfg{/*max_new=*/4, /*steps=*/8, /*block_len=*/0, /*topk=*/-1};
        GenerateResult r = generate(fake, PROMPT, cfg);  // default ReviseConfig: disabled
        for (const Event& e : r.events) assert(!std::holds_alternative<TokensRevised>(e));
        assert((r.generated == std::vector<int>{5, 6, 7, 8}));
    }

    // (c) infill revises too — the gap's low-confidence fills can be reconsidered with both sides.
    {
        FakeAdapter fake(16, -1, -1);
        const std::vector<int> prefix = {1, 2};
        const std::vector<int> suffix = {9, 10};
        GenerateConfig cfg{/*max_new=*/0, /*steps=*/20, /*block_len=*/0, /*topk=*/-1};
        ReviseConfig revise{true, 1.0, 1};
        GenerateResult r = infill(fake, prefix, suffix, /*gap=*/3, cfg, nullptr, {}, revise);
        int n_revised = 0;
        for (const Event& e : r.events)
            if (std::holds_alternative<TokensRevised>(e)) ++n_revised;
        assert(n_revised > 0);
        assert(std::holds_alternative<GenFinished>(r.events.back()));
        assert(r.reason == "length");  // gap fully filled
    }

    // (d) denoise() fills scattered holes and leaves fixed context untouched — the "revise this
    // selection" primitive. board has masks at positions 2, 4, 5; everything else is fixed (7).
    {
        FakeAdapter fake(16, -1, -1);
        const int mask = fake.config().mask_token_id;  // 15
        std::vector<int> board = {7, 7, mask, 7, mask, mask, 7};
        GenerateConfig cfg{/*max_new=*/0, /*steps=*/8, /*block_len=*/0, /*topk=*/-1};
        GenerateResult r = denoise(fake, board, cfg);
        // FakeAdapter argmaxes position p to f(p) = p % 15, so the holes resolve to their own index.
        assert((r.board == std::vector<int>{7, 7, 2, 7, 4, 5, 7}));  // fixed slots untouched, holes filled
        assert(r.reason == "length");
        assert(r.new_tokens == 3);
        assert(std::holds_alternative<GenFinished>(r.events.back()));
    }

    // (e) denoise() with the remask reviser still fills and terminates (holes can be reconsidered).
    {
        FakeAdapter fake(16, -1, -1);
        const int mask = fake.config().mask_token_id;
        std::vector<int> board = {1, mask, mask, 4};
        GenerateConfig cfg{/*max_new=*/0, /*steps=*/20, /*block_len=*/0, /*topk=*/-1};
        GenerateResult r = denoise(fake, board, cfg, nullptr, {}, ReviseConfig{true, 1.0, 1});
        assert((r.board == std::vector<int>{1, 1, 2, 4}));  // pos1->f(1)=1, pos2->f(2)=2
        assert(r.reason == "length");
        int n_revised = 0;
        for (const Event& e : r.events)
            if (std::holds_alternative<TokensRevised>(e)) ++n_revised;
        assert(n_revised > 0);
    }

    std::printf("test_revision: all assertions passed\n");
    return 0;
}

File: engine/lab/tests/test_revision.py
"""remask_lowconf revision policy (DESIGN §5.2) — the "model changes its mind" feature.

Torch-free: the RemaskLowConf policy is pure logic, and the end-to-end path runs on
the FakeAdapter oracle. Verifies revisions fire, respect the per-position cap (so the
loop terminates), stay confined to the generated region, and are strictly opt-in.
"""

import pytest

from cloze_lab.generate import GenerateConfig, generate
from cloze_lab.models.fake import FakeAdapter
from cloze_lab.scheduler.events import GenFinished, TokensRevised
from cloze_lab.scheduler.policies import Candidate, RemaskLowConf, RevisionPolicy, StepContext


class TestRemaskLowConfPolicy:
    def test_is_a_revision_policy(self) -> None:
        assert isinstance(RemaskLowConf(tau_revise=0.5), RevisionPolicy)

    def test_picks_below_tau_only(self) -> None:
        pol = RemaskLowConf(tau_revise=0.5)
        committed = [
            Candidate(pos=2, token_id=9, confidence=0.30),
            Candidate(pos=5, token_id=4, confidence=0.80),  # above tau — keep
            Candidate(pos=7, token_id=1, confidence=0.10),
        ]
        out = pol.revisions(committed, StepContext(step=0), {})
        assert [c.pos for c in out] == [2, 7]  # below tau, pos-ascending

    def test_respects_per_position_cap(self) -> None:
        pol = RemaskLowConf(tau_revise=0.5, max_revisions=2)
        committed = [Candidate(pos=2, token_id=9, confidence=0.3), Candidate(pos=7, token_id=1, confidence=0.1)]
        # pos 7 already hit the cap -> excluded; pos 2 still eligible.
        out = pol.revisions(committed, StepContext(step=0), {7: 2})
        assert [c.pos for c in out] == [2]

    def test_validates_params(self) -> None:
        with pytest.raises(ValueError):
            RemaskLowConf(tau_revise=1.5)
        with pytest.raises(ValueError):
            RemaskLowConf(tau_revise=0.5, max_revisions=0)


class TestRevisionEndToEnd:
    def test_revisions_fire_respect_cap_and_terminate(self) -> None:
        adapter = FakeAdapter(seed=7)
        prompt = adapter.encode("hello cloze")
        cfg = GenerateConfig(max_new=4, steps=12)
        # tau_revise=1.0 re-masks every committed token once (cap=1): a hard stress
        # test that the cap guarantees termination despite aggressive revision.
        result = generate(adapter, prompt, cfg, reviser=RemaskLowConf(tau_revise=1.0, max_revisions=1))

        revised = [e for e in result.events if isinstance(e, TokensRevised)]
        assert revised, "expected at least one revision"

        mask = adapter.config.mask_token_id
        p = len(prompt)
        per_pos: dict[int, int] = {}
        for ev in revised:
            for it in ev.items:
                assert it.old != mask  # a real, previously-committed token was retracted
                assert it.conf < 1.0  # below tau_revise
                assert p <= it.pos < p + cfg.max_new  # within the generated region only
                per_pos[it.pos] = per_pos.get(it.pos, 0) + 1
        assert all(v <= 1 for v in per_pos.values())  # max_revisions cap honored
        assert isinstance(result.events[-1], GenFinished)  # it terminated

    def test_reviser_is_opt_in(self) -> None:
        adapter = FakeAdapter(seed=7)
        prompt = adapter.encode("hello cloze")
        cfg = GenerateConfig(max_new=4, steps=8)
        # No reviser -> no revision events, byte-identical commit path (the goldens
        # are pinned to exactly this).
        result = generate(adapter, prompt, cfg)
        assert not [e for e in result.events if isinstance(e, TokensRevised)]
